### PR TITLE
Adding roles and CRUD controllers

### LINKED ISSUE #5: Role-based model implementation
- [x] Creating CRUD controllers for users
- [x] Connecting Firebase Admin
- [x] Sign Up / Sign In logic
- [x] Roles permissions

### LINKED ISSUE #9: Code base restructuring
- [x] Code base restructuring

### PR DESCRIPTION
**Pull request by vketrys**
- [x] Firebase admin connecting
- [x] Role-based model implementation
- [x] Creating CRUD controllers

### CODE REVIEW

File: .eslintrc [hunk 1]
```diff
@@ -77,5 +77,7 @@
     "prefer-const": [2, { "ignoreReadBeforeAssign": true }],
     "rest-spread-spacing": 2,
     "template-curly-spacing": 2,
+    "arrow-body-style": ["error", "as-needed"],
+    "no-mixed-spaces-and-tabs": 0
```
Comment (esnts) [inline .eslintrc:82]: Why?

File: package.json [hunk 2]
```diff
@@ -39,9 +40,11 @@
     "typescript": "4.9.4"
   },
   "dependencies": {
-    "dotenv": "^16.0.3",
```
Comment (esnts) [inline package.json:42]: ^

File: src/auth.ts [hunk 3]
```diff
@@ -1,36 +1,40 @@
 import { 
 	getAuth, 
-	createUserWithEmailAndPassword, 
 	signInWithEmailAndPassword,
+	UserCredential,
 } from 'firebase/auth';
+import admin from 'firebase-admin';
 import { Request, Response } from 'express';
+
```
Comment (esnts) [inline src/auth.ts]: Empty line?

File: src/auth.ts [hunk 4]
```diff
@@ -1,36 +1,40 @@
...
-		const { email, password } = req.body;
+		const { displayName, email, password, role } = req.body;
   
-		if (!email || !password) {
-			return res.status(statusCodes.unprocessableEntity).json({
-				email: responses.emailRequired,
-				password: responses.passwordRequired,
-			});
+		if (!displayName || !email || !password || !role) {
+			return res.status(statusCodes.unprocessableEntity).send({ message: responses.missingFields });
```
Comment (esnts) [inline src/auth.ts]: I want to know missing fields names

File: src/auth.ts [hunk 5]
```diff
@@ -1,36 +1,40 @@
...
-				return res.status(statusCodes.created).json(user);
-			});
+
+		const { uid } = await admin.auth().createUser({
+			displayName,
+			password,
+			email,
+		});
+
+		await admin.auth().setCustomUserClaims(uid, { role });
```
Comment (esnts) [inline src/auth.ts]: What if i put 'SUPER_PUPER_USER' as `role`? is it valid value for this field?

File: src/constants/codes.ts [hunk 6]
```diff
@@ -5,8 +5,11 @@ export const errorCodes = {
 
 export const statusCodes = {
 	unprocessableEntity: 422,
+	forbidden: 403,
+	unauthorized: 401,
 	badRequest: 400,
 	OK: 200,
```
Comment (esnts) [inline src/constants/codes.ts]: use camelCase for each prop

File: src/constants/roles.ts [hunk 7]
```diff
@@ -0,0 +1,4 @@
+export const roles = {
+	admin: 'admin',
```
Comment (esnts) [inline src/constants/roles.ts]: create enum

File: src/controllers.ts [hunk 8]
```diff
@@ -0,0 +1,83 @@
+import { Request, Response } from 'express';
+import admin from 'firebase-admin';
+
+import { statusCodes } from './constants/codes.js';
+import { responses } from './constants/responses.js';
+
+export const getAll = async(req: Request, res: Response) => {
+	try {
+		const listUsers = await admin.auth().listUsers();
+		const users = listUsers.users.map(mapUser);
```
Comment (esnts) [inline src/controllers.ts]: 😎

File: src/controllers.ts [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+	}
+};
+
+export const remove = async(req: Request, res: Response) => {
+	try {
+		const { id } = req.params;
+
+		const { email } = await admin.auth().getUser(id);
+
+		await admin.auth().deleteUser(id);
```
Comment (esnts) [inline src/controllers.ts]: I can't remove my own account

File: src/controllers.ts [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+  code: string;
+  message: string;
+}
+
+function handleError(res: Response, err: ErrorType) {
+	return res.status(statusCodes.internalServerError).send({ message: `${err.code} - ${err.message}` });
+}
+
+function mapUser(user: admin.auth.UserRecord) {
+	const customClaims = (user.customClaims || { role: '' }) as { role?: string };
```
Comment (esnts) [inline src/controllers.ts]: reuse enum with roles here as type

File: src/controllers.ts [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+	const role = customClaims.role ? customClaims.role : '';
```
Comment (esnts) [inline src/controllers.ts]: customClaims.role ||  '';

File: src/constants/responses.ts [hunk 12]
```diff
@@ -1,5 +1,11 @@
 export const responses = {
-	emailRequired: 'email is required',
-	passwordRequired: 'password is required',
+	emailRequired: 'Email is required',
+	passwordRequired: 'Password is required',
+	missingFields: 'Missing fields',
+	unauthorized: 'Unauthorized',
+	tokenIssue: 'Something wrong with ID Token',
+	roleIssue: 'Current user dont have role',
+	permissionIssue: 'Current user dont have permissions',
```
Comment (slamser56) [inline src/constants/responses.ts]: typo

File: src/auth.ts [hunk 13]
```diff
@@ -1,36 +1,41 @@
 import {
```
Comment (slamser56) [inline src/auth.ts]: Better to move controller, service and routes to separate file

File: src/functions/isAuthorized.ts [hunk 14]
```diff
@@ -0,0 +1,23 @@
+import { Request, Response } from 'express';
+import { statusCodes } from '../constants/codes.js';
+import { responses } from '../constants/responses.js';
+
+export const isAuthorized = (opts: { hasRole: Array<'admin' | 'user'>, allowSameUser?: boolean }) => 
+	// eslint-disable-next-line @typescript-eslint/ban-types
```
Comment (slamser56) [inline src/functions/isAuthorized.ts]: Do you need to use `@typescript-eslint/ban-types` here?

File: src/functions/isAuthorized.ts [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+	 (req: Request, res: Response, next: Function) => {
+		const { role, uid } = res.locals;
+		const { id } = req.params;
+
+		if (opts.allowSameUser && id && uid === id)
```
Comment (slamser56) [inline src/functions/isAuthorized.ts]: You can use `uid === id` instead of `id && uid === id`

File: src/functions/isAutheticated.ts [hunk 16]
```diff
@@ -0,0 +1,36 @@
+import admin from 'firebase-admin';
+import { Request, Response } from 'express';
+import { responses } from '../constants/responses.js';
+import { statusCodes } from '../constants/codes.js';
+
+// eslint-disable-next-line @typescript-eslint/ban-types
```
Comment (slamser56) [inline src/functions/isAutheticated.ts]: Avoid using ban-types

File: src/controllers.ts [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+		return handleError(res, err);
+	}
+};
+
+interface ErrorType {
+  code: string;
+  message: string;
+}
+
+function handleError(res: Response, err: ErrorType) {
```
Comment (slamser56) [inline src/controllers.ts]: Is it seems like common handler?

File: src/controllers.ts (hunk 9 above)
Comment (slamser56) [reply src/controllers.ts]: @vketrys could you think about it?

**Review by esnts — APPROVED** [review]

**Pull request merged**